tests/nvmeof/test_ceph_nvmeof_gateway.py

```diff
@@ -0,0 +1,173 @@
+"""
+Test suite that verifies the deployment of Ceph NVMeoF Gateway
+ with supported entities like subsystems , etc.,
+
+"""
+import json
+
+from ceph.ceph import Ceph
+from ceph.nvmeof.gateway import Gateway, configure_spdk
+from ceph.nvmeof.initiator import Initiator
+from ceph.parallel import parallel
+from ceph.utils import get_node_by_id
+from tests.rbd.rbd_utils import initial_rbd_config
+from utility.log import Log
+from utility.utils import generate_unique_id, run_fio
+
+LOG = Log(__name__)
+
+
+def configure_subsystems(rbd, pool, gw, config):
+    """Configure Ceph-NVMEoF Subsystems."""
+    sub_nqn = config["nqn"]
+    gw.create_subsystem(sub_nqn, config["serial"])
+    gw.create_listener(sub_nqn, config["listener_port"])
+    gw.add_host(sub_nqn, config["allow_host"])
+    if config.get("bdevs"):
+        name = generate_unique_id(length=4)
+        with parallel() as p:
+            count = config["bdevs"].get("count", 1)
+            size = config["bdevs"].get("size", "1G")
+            # Create image
+            for num in range(count):
+                p.spawn(rbd.create_image, pool, f"{name}-image{num}", size)
+
+        with parallel() as p:
+            # Create block device in gateway
+            for num in range(count):
+                p.spawn(
+                    gw.create_block_device,
+                    f"{name}-bdev{num}",
+                    f"{name}-image{num}",
+                    pool,
+                )
+
+        with parallel() as p:
+            # Add namespace
+            for num in range(count):
+                p.spawn(gw.add_namespace, sub_nqn, f"{name}-bdev{num}")
+
+
+def initiators(ceph_cluster, gateway, config):
+    """Run IOs from NVMe Initiators.
+
+    - Discover NVMe targets
+    - Connect to subsystem
+    - List targets and Run FIO on target devices.
+
+    Args:
+        ceph_cluster: Ceph cluster
+        gateway: Ceph-NVMeoF Gateway.
+        config: Initiator config
+
+    Example::
+        config:
+            subnqn: nqn.2016-06.io.spdk:cnode2
+            listener_port: 5002
+            node: node7
+    """
+    client = get_node_by_id(ceph_cluster, config["node"])
+    initiator = Initiator(client)
+    cmd_args = {
+        "transport": "tcp",
+        "traddr": gateway.node.ip_address,
+        "trsvcid": config["listener_port"],
+    }
+    json_format = {"output-format": "json"}
+
+    # Discover the subsystems
```

Comment:
lets have a message when discover is succcessful

Reply:
Updated.